compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -756,6 +756,46 @@ impl<'a> Parser<'a> {
             }
         }
 
+        // We could've continued to parse bounds but we found a type instead of a trait.
+        if !self.can_begin_bound() && self.check_type() {
```

Comment:
```suggestion
        if !self.can_begin_bound() && self.token.can_begin_type() {
```

I recommend you to use `can_begin_type()` over `check_type` since the latter populates the list of *expected tokens* which is not correct strictly speaking (this is why you see the stderr changes in `src/test/ui/fn/fn-recover-return-sign2.rs` and `src/test/ui/parser/trait-object-delimiters.rs`).

Reply:
I was wondering about those unrelated UI test changes. I've made the change, thanks!